drivers: sai: Fix compile time error with clang

dbaluta opened the pull request:
After commit 524b72ce40530 ("toolchain: llvm: Provide working BUILD_ASSERT macro") when compiling with clang we have actual checks and a real assert check using _Static_assert.

Now, when compiling with clang (used by Xtensa internal toolchain) we get the following error.

$ zephyr/drivers/dai/nxp/sai/sai.c:968:29: error: static_assert expression
 is not an integral constant expression

We get this in asserts like this:

BUILD_ASSERT(SAI_DLINE_COUNT(inst) != -1, "...").

This expands to (reduced the macro to easier understand the context):

_Static_assert(((((((I2S_Type *)(uintptr_t)(1493499904U))) == ((I2S_Type *)(0x59040000u)))

... and clang complains that this is not a constant expression.

So, in order to fix the compile time error remove the compile time asserts and replace them with runtime checks.


Fixes: commit 524b72ce40530 ("toolchain: llvm: Provide working BUILD_ASSERT macro")

dbaluta committed:
drivers: sai: Fix compile time error with clang

After commit 524b72ce40530 ("toolchain: llvm: Provide working
BUILD_ASSERT macro") when compiling with clang we have actual
checks and a real assert check using _Static_assert.

Now, when compiling with clang (used by Xtensa internal toolchain)
we get the following error.

$ zephyr/drivers/dai/nxp/sai/sai.c:968:29: error: static_assert expression
 is not an integral constant expression

We get this in asserts like this:

BUILD_ASSERT(SAI_DLINE_COUNT(inst) != -1, "...").

This expands to (reduced the macro to easier understand the context):

_Static_assert(((((((I2S_Type *)(uintptr_t)(1493499904U))) ==
((I2S_Type *)(0x59040000u)))

... and clang complains that this is not a constant expression.

So, in order to fix the compile time error remove the compile time
asserts and replace them with runtime checks.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>
Fixes: commit 524b72ce40530 ("toolchain: llvm: Provide working
BUILD_ASSERT macro")

LaurentiuM1234 commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -886,6 +886,21 @@ static int sai_init(const struct device *dev)
+		return -EINVAL;
+	}
+
+	if (cfg->tx_dline >= cfg->dline_count) {
```
how about:

```
#define SAI_DLINE_COUNT(base) FSL_FEATURE_SAI_CHANNEL_COUNTn(UINT_TO_I2S(base))
```

and

```
if (cfg->tx_dline >= SAI_DLINE_COUNT(cfg->regmap_phys)) {
    /* your code goes here */
}
```

I would like to avoid wasting memory for a variable that's only used in validation.

I also think @decsny's suggestion would work just fine, but then we'd end up with a new DT property whose sole purpose would be to help validate the values passed through the `tx_dataline` and `rx_dataline` DT properties.

LaurentiuM1234 commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -886,6 +886,21 @@ static int sai_init(const struct device *dev)
 	device_map(&data->regmap, cfg->regmap_phys, cfg->regmap_size, K_MEM_CACHE_NONE);
 
+	if (cfg->dline_count <= 0) {
+		LOG_ERR("Invalid number of data lines");
```
suggestion: "bad or unsupported SAI instance". Seems a bit more descriptive?

LaurentiuM1234 commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -886,6 +886,21 @@ static int sai_init(const struct device *dev)
+	}
+
+	if (cfg->tx_dline >= cfg->dline_count) {
+		LOG_ERR("Invalid TX data line index");
```
nit: no capital letter (i.e. "invalid" vs "Invalid") to keep consistent with the other log messages

LaurentiuM1234 commented on drivers/dai/nxp/sai/sai.c:
```diff
@@ -886,6 +886,21 @@ static int sai_init(const struct device *dev)
+	}
+
+	if (cfg->rx_dline >= cfg->dline_count) {
+		LOG_ERR("Invalid RX data line index");
```
nit: no capital letter (i.e. "invalid" vs "Invalid") to keep consistent with the other log messages

LaurentiuM1234 reviewed (commented):
Some comments/nits/suggestions below.

Also, the CI doesn't seem to like the commit message?

dbaluta commented:
Changes since v2: (as per @LaurentiuM1234  comments)
- removed dline_count variable and replaced it with direct call to SAI_DLINE_COUNT
- update error message.

@decsny Your suggestion is OK but I prefer @LaurentiuM1234 because introducing a new device tree property will require to change all existing dts files. 

Also, the other macros are working just fine because they expand to a constant values according to clang. This current macro I removed are not expanding to constants (e.g comparing pointer casts of constants value it's not a constant according to clang implementation).
